### ISSUE TITLE
Why the need for Fast* layers?

### ISSUE THREAD

**Issue by samuela** [SciML/DiffEqFlux.jl#328]
Why the need for `FastChain`, `FastDense`, etc? What's the difference between them and the standard Flux layers?

**Comment by samuela** [comment]
Update: found https://diffeqflux.sciml.ai/dev/FastChain/. It's still not clear to me why the Flux layers are slow, however...

**Comment by ChrisRackauckas** [comment]
Zygote used to have a very large performance penalty for using the `restructure`/`destructure` form that is required to allow for an adjoint ODE, but in the last Zygote release that has largely been fixed. So now there isn't an advantage on CPUs. On GPUs you do get an advantage because the GC is more efficient due to early freeing:

https://github.com/SciML/DiffEqFlux.jl/blob/master/src/fast_layers.jl#L58-L94

but that can probably be removed soon too due to XLA.jl. It needs a bit more testing, but we might just be able to deprecate these very soon.

Other than performance, they are still nice as versions that utilize an explicit parameter vector, though I think that something that uses ComponentArrays.jl is probably a better way forward for that.

**Comment by ChrisRackauckas** [comment]
This is where it came from: https://github.com/SciML/DiffEqFlux.jl/pull/126 but the benchmarks are out of date now due to getting some upstream fixes.

**Comment by ali-ramadhan** [comment]
Updated your script from #126 and ran it on my laptop. Flux layers are faster now but fast layers are still faster by a factor of ~2x. Not sure if a factor of 2 is worth the extra complexity of fast layers but I suppose the user always has the choice.

```julia
using DiffEqFlux, DifferentialEquations, Optim, Flux, Zygote, Test, BenchmarkTools

u0 = Float32[2.; 0.]
datasize = 30
tspan = (0.0f0,1.5f0)

function trueODEfunc(du,u,p,t)
    true_A = [-0.1 2.0; -2.0 -0.1]
    du .= ((u.^3)'true_A)'
end
t = range(tspan[1],tspan[2],length=datasize)
prob = ODEProblem(trueODEfunc,u0,tspan)
ode_data = Array(solve(prob,Tsit5(),saveat=t))

fastdudt2 = FastChain((x,p) -> x.^3,
            FastDense(2,50,tanh),
            FastDense(50,2))

fast_n_ode = NeuralODE(fastdudt2,tspan,Tsit5(),saveat=t)

function fast_predict_n_ode(p)
  fast_n_ode(u0,p)
end

function fast_loss_n_ode(p)
    pred = fast_predict_n_ode(p)
    loss = sum(abs2,ode_data .- pred)
    loss,pred
end

dudt2 = Chain((x) -> x.^3,
             Dense(2,50,tanh),
             Dense(50,2))
n_ode = NeuralODE(dudt2,tspan,Tsit5(),saveat=t)

function predict_n_ode(p)
  n_ode(u0,p)
end

function loss_n_ode(p)
    pred = predict_n_ode(p)
    loss = sum(abs2,ode_data .- pred)
    loss,pred
end

p = fast_n_ode.p
_p,re = Flux.destructure(dudt2)

@test fastdudt2(ones(2),_p) ≈ dudt2(ones(2))
@test fast_loss_n_ode(p)[1] ≈ loss_n_ode(p)[1]
@test Zygote.gradient((p)->fast_loss_n_ode(p)[1], p)[1] ≈ Zygote.gradient((p)->loss_n_ode(p)[1], p)[1]

@btime Zygote.gradient((p)->fast_loss_n_ode(p)[1], p)
@btime Zygote.gradient((p)->fast_loss_n_ode(p)[1], p)
@btime Zygote.gradient((p)->loss_n_ode(p)[1], p)
@btime Zygote.gradient((p)->loss_n_ode(p)[1], p)

@benchmark Zygote.gradient((p)->fast_loss_n_ode(p)[1], p)
@benchmark Zygote.gradient((p)->loss_n_ode(p)[1], p)
```

```
  32.908 ms (117825 allocations: 13.41 MiB)
  32.898 ms (117825 allocations: 13.41 MiB)
  58.723 ms (175614 allocations: 12.22 MiB)
  58.383 ms (175614 allocations: 12.22 MiB)

BenchmarkTools.Trial: 
  memory estimate:  13.41 MiB
  allocs estimate:  117825
  --------------
  minimum time:     32.449 ms (0.00% GC)
  median time:      38.428 ms (0.00% GC)
  mean time:        40.795 ms (5.01% GC)
  maximum time:     69.708 ms (0.00% GC)
  --------------
  samples:          123
  evals/sample:     1

BenchmarkTools.Trial: 
  memory estimate:  12.22 MiB
  allocs estimate:  175614
  --------------
  minimum time:     56.124 ms (0.00% GC)
  median time:      75.526 ms (0.00% GC)
  mean time:        76.051 ms (2.18% GC)
  maximum time:     104.338 ms (0.00% GC)
  --------------
  samples:          66
  evals/sample:     1
```

**Comment by samuela** [comment]
I'd say that 2x is still quite a meaningful perf improvement, but ultimately that's up to @ChrisRackauckas.

**Comment by ali-ramadhan** [comment]
Yeah for sure. I ended up getting rid of my fast layers in favor of Flux layers since I mix `Flux.train!` with `DiffEqFlux.sciml_train!` (probably `GalacticOptim` now) and might want to use Flux's convolutional layers so a 2x slowdown seems worth it for the flexibility.

**Comment by ChrisRackauckas** [comment]
Deprecated as they became an entire library https://github.com/avik-pal/Lux.jl

Flux also got performance improvements, but Lux will be faster if extensive use of `destructure`/`restructure` is necessary. Lux follows the FastChain design of having explicit parameters, so it's directly compatible with things that want the parameters to be an array. Cheers!
